server/index.js

```diff
@@ -54,6 +54,114 @@ app.get('/api/products/:productId', (req, res, next) => {
     .catch(err => next(err));
 });
 
+app.get('/api/cart', (req, res, next) => {
+  const cartId = req.session.cartId;
+  if (!req.session.cartId) {
+    return [];
+  } else {
+    const sql = `select "c"."cartItemId",
+                "c"."price",
+                "p"."productId",
+                "p"."image",
+                "p"."name",
+                "p"."shortDescription"
+            from "cartItems" as "c"
+            join "products" as "p" using ("productId")
+          where "c"."cartId" = $1
+          `;
+
+    const values = [cartId];
+
+    db.query(sql, values)
+      .then(result => res.status(200).json(result.rows));
+  }
+});
+
+app.post('/api/cart', (req, res, next) => {
+  const productId = req.body.productId;
+
+  if (!parseInt(productId, 10)) {
+    return res.status(400).json({ error: 'invalid productId' });
+  }
+
+  const sql = `
+    select "price"
+      from "products"
+    where "productId" = $1
+    `;
+
+  const values = [productId];
+
+  db.query(sql, values)
+    .then(result => {
+      const price = result.rows[0].price;
+      const sql = `
+        insert into "carts" ("cartId", "createdAt")
+        values (default, default)
+        returning "cartId"
+      `;
+      if (!result.rows.length === 0) {
+        return (next(new ClientError(`cannot ${req.method} ${req.originalUrl}`, 404)));
```

Comment:
We need to `throw` a `new ClientError` (instead of passing it to `next()`) in this case, because calling `next()` does not break us out of the promise chain. Otherwise, the promise chain will continue and likely crash the server